app/maptools/measurementmaptool.h

```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Adds point to the end of the recorded geometry; updates recordedGeometry afterwards
+     * Passed point needs to be in map CRS
+     */
+    Q_INVOKABLE void addPoint();
+
+    /**
+     *  Removes last point from recorded geometry if there is at least one point
+     *  Updates recordedGeometry afterwards
+     */
+    Q_INVOKABLE void removePoint();
+
+    /**
+     * Finalizes measurement by forming a polygon if "Close shape" button was clicked
+     * and there are at least 3 points; otherwise, if "Done" button was clicked, forms a polyline.
+     * Updates recorded geometry, calculates perimeter, and calculates the area if it's a polygon.
+    */
+    Q_INVOKABLE void finalizeMeasurement( bool closeShapeClicked );
+
+    /**
+     * Repeats measurement process.
+     * Clears all recorded points and rebuilds the geometry.
+     */
+    Q_INVOKABLE void resetMeasurement();
+
+    // Getter and Setters
+    double lengthWithGuideline() const;
+    void setLengthWithGuideline( const double &length );
+
+    double perimeter() const;
+    void setPerimeter( const double &perimeter );
+
+    double area() const;
+    void setArea( const double &area );
+
+    QPointF crosshairPoint() const;
+    void setCrosshairPoint( const QPointF &point );
+
+    bool canUndo() const;
+    void setCanUndo( bool newCanUndo );
+
+    bool canCloseShape() const;
+    void setCanCloseShape( bool newCanCloseShape );
+
+    bool isValidGeometry() const;
+    void setIsValidGeometry( bool hasValidGeometry );
+
+    bool measurementFinalized() const;
+    void setMeasurementFinalized( bool newMeasurementFinalized );
+
+    const QgsGeometry &recordedGeometry() const;
+    void setRecordedGeometry( const QgsGeometry &newRecordedGeometry );
+
+    QgsGeometry existingVertices() const;
+    void setExistingVertices( const QgsGeometry &vertices );
+
+    void resetMapSettings();
+    void updateMapSettings( InputMapSettings *newMapSettings );
+
+  signals:
+    void lengthWithGuidelineChanged( const double &lengthWithGuideline );
+    void perimeterChanged( const double &perimeter );
+    void areaChanged( const double &area );
+    void canUndoChanged( bool canUndo );
+    void canCloseShapeChanged( bool canUndo );
+    void measurementFinalizedChanged( bool measurementFinalized );
+    void recordedGeometryChanged( const QgsGeometry &recordedGeometry );
+    void existingVerticesChanged( const QgsGeometry &vertices );
+    void crosshairPointChanged( const QPointF &crosshairPoint );
+    void isValidGeometryChanged( bool canFinalize );
+
+  protected:
+    void rebuildGeometry();
+    void checkCanCloseShape();
+
+  public slots:
+    void updateDistance();
+
+  private:
+    QVector<QgsPoint> mPoints;
+    QgsGeometry mRecordedGeometry;
+    QgsGeometry mExistingVertices;
+    QgsDistanceArea mDistanceArea;
+    QPointF mCrosshairPoint;
+    double mLengthWithGuideline = 0;
+    double mPerimeter = 0;
+    double mArea = 0;
```

Comment:
this feels unnecessary complicated to me:
- Why do we need 3 sets of points (`mPoints`, `mRecordedGeometry`, `mExistingVertices`)? Can't we just have one?
- Why do we need to calculate and store the derived attributes (`mArea`, `mPerimeter`, `mLengthWithGuideline`)? Can't we directly fetch them from the `QgsGeometry` methods? I'm afraid our stored values could eventually get out of sync with the underlying geometry